Update rust reference info about closures

roxelo opened the pull request:
cc @arora-aman

workingjubilee committed:
Add RangeFrom patterns

workingjubilee committed:
More formal writing and links

workingjubilee committed:
Offer example

ehuss committed:
Expand on Unicode identifiers.

MarcusDunn committed:
added an example for binding after @

MarcusDunn committed:
removed repeated word

MarcusDunn committed:
added a comma

kklibo committed:
minor grammar and typo fixes

kklibo committed:
improve for loop description (copied from 'while let' equivalent)

kklibo committed:
clean up grammar in 'slice types'

kklibo committed:
disambiguate sentence

The phrase 'may not' could mean 'might not' or 'is not allowed to'.

kklibo committed:
disambiguate sentence

The original sentence could mean that "its destructor is run" and "it is dropped" are two separate possibilities, rather than two names for the same action.

arora-aman commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
-}
+### Capturing references in move contexts
+
+Rust doesn't allow moving fields out references as a result in the case of move closures, when the closure takes ownership of a path, it will only take ownership of a prefix of path that runs up until the first dereference (this may be the entire path if there are no dereferences in the path).
```
moving fields out of

arora-aman replied:
Instead of truncate ... we reborrow the data being accessed throw the reference

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
 ## Capture modes
 
-The compiler prefers to capture a closed-over variable by immutable borrow,
+The compiler prefers to capture a value basedimmutable borrow,
```
```suggestion
The compiler prefers to capture a value by immutable borrow,
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
-usually used to allow the closure to outlive the captured values, such as if the
-closure is being returned or used to spawn a new thread.
+## Capture Precision
+The precise path that gets captured is typically the full path that is used in the closure, but there are cases where we will only capture a prefix of the path.
```
```suggestion

The precise path that gets captured is typically the full path that is used in the closure, but there are cases where we will only capture a prefix of the path.
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+
+### Shared prefix
+
+In the case where a path and one of the ancestor’s of that path are both captured by a closure, the ancestor path is captured with `CaptureMode = max(AncestorCaptureMode, DescendantCaptureMode)` using the strict weak ordering 
```
```suggestion
In the case where a path and one of the ancestor’s of that path are both captured by a closure, the ancestor path is captured with the highest capture mode among the two captures,`CaptureMode = max(AncestorCaptureMode, DescendantCaptureMode)`, using the strict weak ordering 
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
-}
+### Capturing references in move contexts
+
+Rust doesn't allow moving fields out references as a result in the case of move closures, when the closure takes ownership of a path, it will only take ownership of a prefix of path that runs up until the first dereference (this may be the entire path if there are no dereferences in the path).
```
```suggestion
Rust doesn't allow moving fields out of references. As a result, in the case of move closures, when a closure takes ownership of a path, it will only take ownership of a prefix of the path that runs up until the first dereference (this may be the entire path if there are no dereferences in the path).
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
-to capture `self` by mutable reference. But since `self.set` is already
-borrowed to iterate over, the code would not compile.
+### Raw pointer dereference
+In Rust it's `unsafe` to dereference a raw pointer therefore closures will only capture the prefix of a path that runs upto but not including the first dereference of a raw pointer.
```
```suggestion
In Rust, it's `unsafe` to dereference a raw pointer. Therefore, closures will only capture the prefix of a path that runs up to, but not including, the first dereference of a raw pointer.
```

roxelo commented on src/types/closure.md:
````diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+```rust,
+struct T(String, String);
+
+let foo = T(String::from("foo"), String::from("bar"));
````
```suggestion
let t = T(String::from("foo"), String::from("bar"));
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+struct T(String, String);
+
+let foo = T(String::from("foo"), String::from("bar"));
+let foo = &foo as *const T;
```
```suggestion
let t = &t as *const T;
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+let foo = &foo as *const T;
+
+let c = || unsafe {
+    println!("{}", (*foo).0); // closure captures t
```
```suggestion
    println!("{}", (*t).0); // closure captures t
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+
+### Reference into unaligned `struct`s
+
+In Rust it's `unsafe` to hold references to unaligned fields in a structure and therefore closures will only capture the prefix of the path that runs upto but not including the first field access into an unaligned struct.
```
```suggestion
In Rust, it's `unsafe` to hold references to unaligned fields in a structure, and therefore, closures will only capture the prefix of the path that runs up to, but not including, the first field access into an unaligned structure.
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+
+### `Box` vs other `Deref` implementations
+
+The compiler special cases the implementation of the Deref trait for `Box` and the feature treats `Box` as a special entity as well.
```
```suggestion
The compiler treats the implementation of the Deref trait for `Box` differently, as it is considered a special entity.
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+
+The compiler special cases the implementation of the Deref trait for `Box` and the feature treats `Box` as a special entity as well.
+
+To draw a comparison we look at explamples involving `Rc` and `Box`. The `*rc` is desugared to a call to the trait method `deref` method defined on `Rc`, but `*box` is special cased by the compiler. This implies that the compiler is able to do precise capture on contents of the Box.
```
```suggestion
For example, let us look at examples involving `Rc` and `Box`. The `*rc` is desugared to a call to the trait method `deref` defined on `Rc`, but since `*box` is treated differently by the compiler, the compiler is able to do precise capture on contents of the `Box`.
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+
+#### Non `move` closure
+
+In a closure that isn't `move`, if the contents of the box are not moved into the closure body the contents of the box are precisely captured.
```
```suggestion
In a non `move` closure, if the contents of the `Box` are not moved into the closure body, the contents of the `Box` are precisely captured.
```

roxelo commented on src/types/closure.md:
````diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+};
+```
+
+However if the contents of the box are moved into the closure then the box is entierly captured. This is done to do so the amount of data that needs to be moved into the closure is minimized.
````
```suggestion
However, if the contents of the `Box` are moved into the closure, then the box is entirely captured. This is done so the amount of data that needs to be moved into the closure is minimized.
```

roxelo commented on src/types/closure.md:
```diff
@@ -42,48 +50,148 @@ so that the call to `f` works as if it were:
+
+#### `move` closure
+
+Similar to moving contents of a box in a non-`move` closure reading the contents of a box in a `move` closure results in capturing the box entierly.
```
```suggestion
Similarly to moving contents of a `Box` in a non-`move` closure, reading the contents of a `Box` in a `move` closure will capture the `Box` entirely.
```

roxelo commented on src/types/closure.md:
```diff
@@ -178,3 +288,117 @@ Because captures are often by reference, the following general rules arise:
+
+## Drop Order
+
+If a closure captures a field of a composite types such as structs, tuples, and enums by ByValue, the field's lifetime would now be tied to the closure. As a result, it is possible for disjoint fields of a composite types to be dropped at different times.
```
```suggestion
If a closure captures a field of a composite types such as structs, tuples, and enums by value, the field's lifetime would now be tied to the closure. As a result, it is possible for disjoint fields of a composite types to be dropped at different times.
```

roxelo commented on src/types/closure.md:
````diff
@@ -178,3 +288,117 @@ Because captures are often by reference, the following general rules arise:
+```
+
+## Wild Card Patterns
+Closures now only capture data that needs to be read, which means the following closures will not capture `x`
````
```suggestion
Closures only capture data that needs to be read, which means the following closures will not capture `x`
```

roxelo commented on src/types/closure.md:
```diff
@@ -178,3 +288,117 @@ Because captures are often by reference, the following general rules arise:
+
+## Closure types difference
+
+In Edition 2018 and before, a closure would capture variables in its entirety. This meant that for the example used in the [Closure types]() section the generated closure type would look something like this:
```
```suggestion
In Edition 2018 and before, a closure would capture variables in its entirety. This means that for the example used in the [Closure types] section, the generated closure type would instead look something like this:
```

arora-aman replied:
we should create a link above and link them here? I think if look at a rust RFC there should be expamples for it.

I'm refering to the fact that [Clsoure types] doesn't have a link associated with it

roxelo commented on src/types/closure.md:
````diff
@@ -178,3 +288,117 @@ Because captures are often by reference, the following general rules arise:
+    }
+}
+```
+and the call to `f` would work as if it were:
````
```suggestion
and the call to `f` would work as follows:
```

roxelo commented on src/types/closure.md:
```diff
@@ -178,3 +288,117 @@ Because captures are often by reference, the following general rules arise:
+## Capture precision difference
+
+Composite types such as structs, tuples, and enums are always captured in its intirety,
+not by individual fields. It may be necessary to borrow into a local variable in order to capture a single field:
```
```suggestion
not by individual fields. As a result, it may be necessary to borrow into a local variable in order to capture a single field:
```

roxelo commented on src/types/closure.md:
```diff
@@ -178,3 +288,117 @@ Because captures are often by reference, the following general rules arise:
+
+Regardless of if the data will be read by the closure, i.e. in case of wild card patterns, if a variable defined outside the closure is mentioned within the closure the variable will be captured in its entirety.
+
+## Other traits difference
```
I think we can actually delete this section. This is because the modified rules stated "the captured values" and we already clarified in the earlier Edition2018 subsection that the values captured are variables that are captured in its entirety

arora-aman replied:
I'm fine removing it

roxelo commented on src/types/closure.md:
```diff
@@ -178,3 +288,117 @@ Because captures are often by reference, the following general rules arise:
+
+## Drop order difference
+
+As composite types are now fully captured, a closure which captures one of those composite types ByValue would drop it as the same time as closure gets dropped.
```
```suggestion
As composite types are captured in their entirety, a closure which captures one of those composite types by value would drop the captured variable at the same time as closure gets dropped.
```

arora-aman replied:
would drop the entire captured variable

roxelo reviewed (commented):
@arora-aman I think there are a few things we want to change. Mainly wording

arora-aman commented on src/types/closure.md:
````diff
@@ -178,3 +288,117 @@ Because captures are often by reference, the following general rules arise:
+} // tuple.1 dropped here -----------------------------+
+```
+
+## Wild Card Patterns
````
this section is already part of the Capture precision above

roxelo committed:
Update closure types documentation so it includes information about RFC2229

roxelo committed:
Address comments

arora-aman committed:
Copy capture analyis alogrithm from hackmd

arora-aman committed:
Update copy type, optimization, make algorithm (place, mode)

roxelo committed:
fix stray merge conflict

carols10cents committed:
Remove incorrect apostrophe

This should be the other form of its.

ehuss committed:
Merge pull request #1076 from carols10cents/patch-1

Remove incorrect apostrophe

ehuss committed:
Merge pull request #1022 from ehuss/ident-unicode

Expand on Unicode identifiers.

ehuss committed:
Revert phrase change for destructors.

It is not clear if the original intended to mean these as two separate
concepts, or synonyms for the same thing.

ehuss committed:
Merge pull request #1037 from kklibo/fix_typos

Fix typos + grammar

JohnTitor committed:
Mention "function item type to `fn pointer`" coercion

Commit:
array-expr.md: Fix typo;'polished' sentence

ehuss committed:
Merge pull request #1080 from daerich/array-typo

array-expr.md: fix typo; 'polish' sentence

ehuss committed:
Merge pull request #1027 from MarcusDunn/master

example for bindings after at

ehuss committed:
Merge pull request #1079 from JohnTitor/mention-one-more-coercion

Mention "function item type to `fn pointer`" coercion

arora-aman committed:
Minor edits from Feedback

arora-aman committed:
Add details on truncation and ref-uniq

workingjubilee committed:
add spacing in src/patterns.md

Co-authored-by: Eric Huss <eric@huss.org>

fee1-dead committed:
Allow users to change status labels

ehuss committed:
Merge pull request #1083 from fee1-dead/master

Allow users to change status labels

Commit:
expressions.md: Attempt fixing broken grammar in Mutability paragraph

JohnTitor committed:
Merge pull request #1084 from daerich/mutctx_typo

expressions.md: Attempt fixing broken grammar in Mutability paragraph

ehuss committed:
Fix recursion_limit example.

workingjubilee committed:
Update with slice pattern restrictions

ehuss committed:
Merge pull request #900 from workingjubilee/rangefrom-pat

Document RangeFrom patterns

arora-aman committed:
Update closure algorithm per #88467 #88477

JohnTitor committed:
Merge pull request #1087 from ehuss/fix-recursion_limit-example

LazaroHurtado committed:
fixed link typo

ehuss committed:
Merge pull request #1094 from LazaroHurtado/master

Fixed link typo in Generics.md

waywardmonkeys committed:
Use subtrait/supertrait, not sub-trait, super-trait.

There are definitions of subtrait and supertrait, so use the
official terminology.

ehuss committed:
Merge pull request #1095 from waywardmonkeys/consistent-supertrait-subtrait

Use subtrait/supertrait, not sub-trait, super-trait.

arora-aman committed:
Update src/types/closure.md

Co-authored-by: Josh Triplett <josh@joshtriplett.org>
